benchmark/atoi.cpp

```diff
@@ -115,6 +123,110 @@ blockAlignedLoad(PtrT *pointerInsideBlock, Block *b) {
     return { base, misalignment };
 }
 
+std::size_t leadingSpacesCount(const char *p) noexcept {
+    using S = swar::SWAR<8, uint64_t>;
+    S bytes;
+    auto [base, misalignment] = blockAlignedLoad(p, &bytes.m_v);
+    auto bitDisplacement = 8 * misalignment;
+
+    // deal with misalignment setting the low part to spaces
+    constexpr static S
+        AllSpaces{meta::BitmaskMaker<uint64_t, ' ', 8>::value},
+        AllOn = ~S{0};
+    // blit the spaces in
+    auto mask = S{AllOn.value() << bitDisplacement};
+    auto misalignedEliminated = bytes & mask;
+    auto spacesIntroduced = AllSpaces & ~mask;
+    bytes = spacesIntroduced | misalignedEliminated;
+    for(;;) {
+        auto spacesThisBlock = leadingSpacesCountAligned(bytes);
+        base += spacesThisBlock;
+        if(8 != spacesThisBlock) { return base - p; }
+        memcpy(&bytes.m_v, base, 8);
+    }
+}
+
+auto leadingDigitsCount(const char *p) noexcept {
+    using S = swar::SWAR<8, uint64_t>;
+    S bytes;
+    auto [base, misalignment] = blockAlignedLoad(p, &bytes.m_v);
+    auto bitDisplacement = 8 * misalignment;
+    constexpr static S
+        AllZeroCharacter{meta::BitmaskMaker<uint64_t, '0', 8>::value},
+        AllOn = ~S{0};
+    // blit the zero-characters to the misaligned part
```

Comment:
If we're going to shift something, and we're counting leading digits, why not shift out the misaligned part?  Load "xx12'34ab" , shift to "1234'abxx"?, get 4?

Reply:
The alternative you suggest is valid but sub-optimal, I actually used that option until I detected the improvement:
In this prolog to the loop, this code only brings the loop invariant in.
The loop invariant:
you work with the whole block, `base + nonDigitIndex - p` is the count
As far as I am able to tell, you'd need a modified version of the loop to calculate the count in a different way and hoist it to the prolog to work as an special first iteration, leading to non-optimal object code: in the code as it is, the extra blitting is better than a modified iteration.
Please check my reasoning, there might be a slight improvement here I may be missing.

Reply:
I think you are right.  We can increment base by the seen number of spaces in the prologue, but then we have to change the exit condition to something else (and I think that 'something else' is more expensive than 'did leading spaces returned eight' as a predicate).  Maybe there is some way, but everything looks like a modulo or a trick.)